src/main/java/org/thoughtcrime/securesms/WebxdcStoreActivity.java

```diff
@@ -0,0 +1,135 @@
+    setContentView(R.layout.web_view_activity);
+    rpc = DcHelper.getRpc(this);
+    dcContext = DcHelper.getContext(this);
+    WebView webView = findViewById(R.id.webview);
+
+    ActionBar actionBar = getSupportActionBar();
+    if (actionBar != null) {
+      actionBar.setDisplayHomeAsUpEnabled(true);
+      actionBar.setTitle(R.string.webxdc_apps);
+    }
+
+    webView.setWebViewClient(new WebViewClient() {
+      @Override
+      public boolean shouldOverrideUrlLoading(WebView view, String url) {
+        view.loadUrl(url);
+        return true;
+      }
+
+      @TargetApi(Build.VERSION_CODES.N)
+      @Override
+      public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
+        view.loadUrl(request.getUrl().toString());
+        return true;
+      }
+
+      @Override
+      public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
+        WebResourceResponse res = interceptRequest(request.getUrl().toString());
+        if (res == null) {
+          res = super.shouldInterceptRequest(view, request);
+        }
+        return res;
+      }
+    });
+
+    WebSettings webSettings = webView.getSettings();
+    webSettings.setJavaScriptEnabled(true);
+    webSettings.setAllowFileAccess(false);
+    webSettings.setAllowContentAccess(false);
+    webSettings.setGeolocationEnabled(false);
+    webSettings.setAllowFileAccessFromFileURLs(false);
+    webSettings.setAllowUniversalAccessFromFileURLs(false);
+    webSettings.setDatabaseEnabled(true);
+    webSettings.setDomStorageEnabled(true);
+    webView.setNetworkAvailable(true); // this does not block network but sets `window.navigator.isOnline` in js land
+
+    webView.loadUrl(Prefs.getWebxdcStoreUrl(this));
+  }
+
+  private WebResourceResponse interceptRequest(String url) {
+    WebResourceResponse res = null;
+    try {
+      if (url == null) {
+        throw new Exception("no url specified");
+      }
+      HttpResponse httpResponse = rpc.getHttpResponse(dcContext.getAccountId(), url);
+      String mimeType = httpResponse.getMimetype();
+      if (mimeType == null) {
+        mimeType = "application/octet-stream";
+      }
+
+      String ext = MediaUtil.getFileExtensionFromUrl(Uri.parse(url).getPath());
+      if ("xdc".equals(ext)) {
+        Intent intent = new Intent();
+        Uri uri = PersistentBlobProvider.getInstance().create(this, httpResponse.getBlob(), mimeType, "app.xdc");
+        intent.setType(mimeType);
+        intent.setData(uri);
+        setResult(Activity.RESULT_OK, intent);
+        finish();
+      } else {
+        ByteArrayInputStream data = new ByteArrayInputStream(httpResponse.getBlob());
+        res = new WebResourceResponse(mimeType, httpResponse.getEncoding(), data);
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+      ByteArrayInputStream data = new ByteArrayInputStream(("Error: " + e.getMessage()).getBytes());
```

Comment:
Should show a nicer error message when there is no internet (or make things work without internet)

Reply:
working without internet is out of the scope of this PR, it would be possible if we used a webxdc app with internet access (which I recommend to do) then the webxdc is offline-first and will have something to load and show if the fetching of the remote metadata failed

Reply:
<img width=320 src=https://github.com/user-attachments/assets/8afad88e-7240-4e12-b47a-0ca37a97d3b0>

yeah, that can be improved, however, for a first, experimental thing this should also be no blocker :) i have seen worse error messages :)

working without internet is out of scope of this PR, iirc, the idea is to let core handle that at some point

Reply:
about  the error, one of the problems is that the RPC returns that string instead of just the message, but yeah all that is out of scope of this PR

Reply:
```suggestion
      ByteArrayInputStream data = new ByteArrayInputStream(("Could not load apps. Are you online?\n\n" + e.getMessage()).getBytes());
```